### ISSUE TITLE
pow(i, 2) returns -1 with small imaginary part

### ISSUE THREAD

**Issue by keyten** [josdejong/mathjs#550]
![Screenshot](http://i.imgur.com/jlVPISG.png)

**Comment by josdejong** [comment]
Hi, this is a round-off error which is "normal" when working with floating point numbers. Google for `0.1+0.2 != 0.3`. The calculator on the website of mathjs rounds the output to a limited number of digits to hide these round-off errors. See also: http://mathjs.org/docs/datatypes/numbers.html#roundoff-errors

**Comment by keyten** [comment]
Hmm... Yep, thanks.
But it looks like unexpected behavior, what about fixing it (for any precision)?
Make `math.config({ floatfix: true })`?

**Comment by josdejong** [comment]
It no *unexpected* behavior: this is a consequence of how floating point numbers work, and there is no fix for this. But it's definitely *unwanted* behavior. 

math.js has support for BigNumbers and Fractions, which help a lot. However, Complex still uses (float) numbers internally, so it will not help when doing complex computations. 

If you find this matter interesting, just google for it, or read the mathjs docs on numbers, BigNumbers, and Fractions.

**Comment by infusion** [comment]
I tried to face this with complex.js, but there is not too much we can do, even with more clever numerical tricks. But I think the problem isn't that big, as long as you use these numbers for further calculations. It might be true, that the error propagates through all calculations, but in practice, it doesn't make a big difference. The other thing is, that it doesn't "look right" or even pretty. If you want to print the number, don't rely on the object attributes but print it with a proper rounding setting, like Jos said.

**Comment by infusion** [comment]
For a complex number `z^n=(a+bi)^n`, I identified some edge cases:

If `n` is a real number, a>=0 and b=0, we can simply calculate Math.pow(a, n).
If `n` is an integer and a=0, we can calculate (bi)^n, splitting n into `mod 4` partitions.

I think with more investigation, we could spot some more such cases. The only question remaining is, if we should tread every of such cases in order to keep accuracy high. If you ask me, it's quite nice to remove this 1e^-15 rubbish from the unit tests. On the other hand, it's the simplicity in our thinking to make these trivial cases look good. Complex numbers have applications in the computation of alternating current, fractals and so on and focusing on making these applications work fine, would be much more precious.

I spend quite some time on the pow function to come up with a simple, fast and numerical stable implementation. At the end, there were only 4 lines of code. Just with these edge cases mentioned above, I added some additional 20 LOC. I don't know if it's worth that. It is becoming more a political question. What do you think Jos?

**Comment by josdejong** [comment]
There are a lots of these cases where you end up with a round-off error. Also a lot of "special" cases like `pow(i, 2)` or `e^(i*pi)` (Euler's identity). We can spend a lot of effort optimizing for a few of these cases, I doubt if that's worth it.

I would rather search in a different direction for a "real" solution. Not so long ago we extended `Unit` such that it can store it's value not only as number, but also accepts BigNumber and Fraction. For Complex we could do exactly the same. For performance reasons you may want to create two different implementations for Complex, a fast one only working with number, and a more generic one accepting different numeric types.

**Comment by infusion** [comment]
Yep, that's what I'm saying. We should focus on real applications and not these special cases which come to mind when writing some easy numbers for unit tests. I don't know if you plan symbolic calculations, like matlab or wolfram is capable to do. If so, adding complex numbers there would be the right direction to fix such numerical errors.

Using BigNumber as a complex.js backend is a good idea. When it comes to fraction as the base of complex numbers, something inside of me rebels. Not only that we must implement trigonometric functions on our own, which deliver a silly rational approximation, it also has no mathematical foundation.

**Comment by josdejong** [comment]
Indeed Fractions cannot be used with all types of operations (like trigonometric), but for all basic arithmetics (add, subtract, multiply, divide, etc) they can be very useful and save you from these round-off errors.

**Comment by Hypercubed** [comment]
@infusion if you do start working on a version of complex.js using BigNumber backed I'll contribute.  I have a small implementation of a BigComplex class already.

**Comment by infusion** [comment]
It's probably not the right topic here, but to outline the current planning (maybe we should talk via email or something else directly): I work on centralizing the complex numbers using my own complex.js implementation. There will be a bigger update on my project as well. For the last days, I tried to squeeze a bit more accuracy out of the pow() function, based on this topic here, but a certain error will remain - which resulted in the answers before.

I do not plan to make my complex implementation modular in order to have different backends, like BigNumber (or even fractions) and I agree with Jos, that a performant implementation should remain. The easiest thing we could do is using my codebase and replace every elementary operation with an abstract type, which implements either normal javascript operations or a more complex type.

This duplicating is quite easy and BigComplex is a name, I like a lot. I wonder, if this approach stays in contrast to the intend of slimming weight off the library, as we must duplicate everything. On the other hand, maintaining two implementations, aiming for the same goal can easily lead to strange errors.

Parsers are always a size blowing element. What we could do is adding some comments to complex.js, fraction.js and bigcomplex.js to mark the start and end of sections which can be removed by the build process of math.js. So the libraries on their own can be used normally and math.js cares about the string handling by itself. The parser then could also decide which class must be used.

**Comment by Hypercubed** [comment]
Maybe exactly what you don't want to do but it would be amazing to have a set of modules that each implement the same api (plus, minus, etc).  I'm thinking something like the fantasy-land spec for mumeric types.  I can imagine "number" (api wrapper around js Number), decimal, complex, fraction/rational, etc.  Something like math.js (using typed-function) could interact with each in a consistant manner, converting between types as needed.  Higher order types, like complex, vector, and matrix, could use either number or decimal as a back end.

Personally I'm interested in creating a Dec64 lib.  Being able to wrap this in a complex type would be amazing.

If you want to have a discussion elsewhere please let me know.

**Comment by josdejong** [comment]
A Dec64 or other numeric type not suffering from round-off errors would be really cool. It would be slower than numbers, but it's probably way faster than using BigNumbers (we would have to validate that of course).

**Comment by infusion** [comment]
I was playing around with an uber-data-type, which combines rational and irrational numbers as well as complex numbers. I developed the idea of the type as: `s * x * n / d`, where s is the sign, x an irrational part, normalized to the interval `[0, 1]` and n and d the numerator and denominator. When we think of x, there are a lot of cases where we can express it better, like with an rational exponent: `s * x^(n'/d') * n / d`. But we also need negative exponents and irrational exponents, the same for the rational part and we end up with something like: `s * x^(s' * x' * n' / d') * (n / d)^(s'' * x'' * n'' / d'')`. The properties of this construct are quite manageable. However, if we duplicate the whole for the imaginary part to bring complex numbers into play, things are getting really thorny if you try to think of adding two of such numbers. Developing a symbolic calculation engine would be easier I think.

**Comment by josdejong** [comment]
You mean some sort of ComplexFraction? hm, I don't know. I would say you just need to store more digits and store the exponent in powers of 10, basically what BigNumber libraries do and what Dec64 proposes.

**Comment by infusion** [comment]
Yep, a ComplexFraction if you will. But as I said, it has really poor properties as well as a lot of edge cases which have to be implemented, even for simple operations. Throwing more bits on the problem is probably the best solution.

**Comment by farzher** [comment]
eval `e^(i*pi)+1` renders as `1.224646799e-16i`

Even this doesn't help `mathjs.format(result, {precision: 10})`
The precision option doesn't work for complex numbers? It works for `.1+.2` How can I get this to render as 0?

**Comment by josdejong** [comment]
What you can do is format using `fixed` notation:

```js
var result = math.eval('e^(i*pi)+1')
math.format(result, {notation: "fixed", precision: 10}) // outputs '0.0000000000i' 
```

To get rid of the trailing zeros you could parse and format the rounded value again:
```js
var result = math.eval('e^(i*pi)+1')
var rounded = math.format(result, {notation: "fixed", precision: 10}) // '0.0000000000i'
var formatted = math.format(math.eval(rounded))   // outputs '0'
```

**Comment by farzher** [comment]
> To get rid of the trailing zeros you could parse and format the rounded value again:
> 
> var result = math.eval('e^(i*pi)+1')
> var rounded = math.format(result, {notation: "fixed", precision: 10}) // '0.0000000000i'
> var formatted = math.format(math.eval(rounded))   // outputs '0'

That worked perfectly, thanks
